File: record_android/android/src/main/java/com/llfbandit/record/methodcall/MethodCallHandlerImpl.java
package com.llfbandit.record.methodcall;

import android.app.Activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.llfbandit.record.Utils;
import com.llfbandit.record.permission.PermissionManager;
import com.llfbandit.record.record.RecordConfig;
import com.llfbandit.record.record.format.AudioFormats;

import java.io.IOException;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class MethodCallHandlerImpl implements MethodCallHandler {
  private final PermissionManager permissionManager;
  private final BinaryMessenger messenger;
  @Nullable
  private Activity activity;

  private final ConcurrentHashMap<String, RecorderWrapper> recorders = new ConcurrentHashMap<>();

  public MethodCallHandlerImpl(
      @NonNull PermissionManager permissionManager,
      @NonNull BinaryMessenger messenger) {

    this.permissionManager = permissionManager;
    this.messenger = messenger;
  }

  public void dispose() {
    for (RecorderWrapper recorder : recorders.values()) {
      recorder.dispose();
    }
    recorders.clear();
  }

  public void setActivity(@Nullable Activity activity) {
    this.activity = activity;

    for (RecorderWrapper recorder : recorders.values()) {
      recorder.setActivity(activity);
    }
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    String recorderId = call.argument("recorderId");
    if (recorderId == null || recorderId.isEmpty()) {
      result.error("record", "Call missing mandatory parameter recorderId.", null);
      return;
    }

    if (call.method.equals("create")) {
      try {
        createRecorder(recorderId);
        result.success(null);
      } catch (Exception e) {
        result.error("record", "Cannot create recording configuration.", e.getMessage());
      }
      return;
    }

    RecorderWrapper recorder = recorders.get(recorderId);
    if (recorder == null) {
      result.error(
          "record",
          "Recorder has not yet been created or has already been disposed.", null
      );
      return;
    }

    switch (call.method) {
      case "start":
        try {
          RecordConfig config = getRecordConfig(call);
          recorder.startRecordingToFile(config, result);
        } catch (IOException e) {
          result.error("record", "Cannot create recording configuration.", e.getMessage());
        }
        break;
      case "startStream":
        try {
          RecordConfig config = getRecordConfig(call);
          recorder.startRecordingToStream(config, result);
        } catch (IOException e) {
          result.error("record", "Cannot create recording configuration.", e.getMessage());
        }
        break;
      case "stop":
        recorder.stop(result);
        break;
      case "pause":
        recorder.pause(result);
        break;
      case "resume":
        recorder.resume(result);
        break;
      case "isPaused":
        recorder.isPaused(result);
        break;
      case "isRecording":
        recorder.isRecording(result);
        break;
      case "hasPermission":
        permissionManager.hasPermission(result::success);
        break;
      case "getAmplitude":
        recorder.getAmplitude(result);
        break;
      case "listInputDevices":
        result.success(null);
        break;
      case "dispose":
        recorder.dispose();
        recorders.remove(recorderId);
        result.success(null);
        break;
      case "isEncoderSupported":
        String codec = call.argument("encoder");
        boolean isSupported = AudioFormats.isEncoderSupported(
            AudioFormats.getMimeType(Objects.requireNonNull(codec))
        );
        result.success(isSupported);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  private void createRecorder(String recorderId) {
    RecorderWrapper recorder = new RecorderWrapper(recorderId, messenger);
    recorder.setActivity(activity);
    recorders.put(recorderId, recorder);
  }

  private RecordConfig getRecordConfig(@NonNull MethodCall call) throws IOException {
    return new RecordConfig(
        call.argument("path"),
        Utils.firstNonNull(call.argument("encoder"), "aacLc"),
        Utils.firstNonNull(call.argument("bitRate"), 128000),
        Utils.firstNonNull(call.argument("samplingRate"), 44100),
        Utils.firstNonNull(call.argument("numChannels"), 2),
        call.argument("device"),
        Utils.firstNonNull(call.argument("autoGain"), false),
        Utils.firstNonNull(call.argument("echoCancel"), false),
        Utils.firstNonNull(call.argument("noiseCancel"), false)
    );
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/methodcall/MethodCallHandlerImpl.kt
package com.llfbandit.record.methodcall

import android.content.Context
import com.llfbandit.record.permission.PermissionManager
import com.llfbandit.record.record.RecordConfig
import com.llfbandit.record.record.device.DeviceUtils
import com.llfbandit.record.record.format.AudioFormats
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import java.util.Objects
import java.util.concurrent.ConcurrentHashMap

class MethodCallHandlerImpl(
  private val permissionManager: PermissionManager,
  private val messenger: BinaryMessenger,
  private val appContext: Context
) : MethodCallHandler {
  private val recorders = ConcurrentHashMap<String, RecorderWrapper>()

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    val recorderId = call.argument<String>("recorderId")

    if (recorderId.isNullOrEmpty()) {
      result.error("record", "Call missing mandatory parameter recorderId.", null)
      return
    }

    if (call.method == "create") {
      createRecorder(recorderId, result)
      return
    }

    val recorder = recorders[recorderId]
    if (recorder == null) {
      result.error(
        "record",
        "Recorder has not yet been created or has already been disposed.", null
      )
      return
    }

    when (call.method) {
      "start" -> recorder.startRecordingToFile(RecordConfig.fromMap(call, appContext), result)
      "startStream" -> recorder.startRecordingToStream(RecordConfig.fromMap(call, appContext), result)
      "stop" -> recorder.stop(result)
      "pause" -> recorder.pause(result)
      "resume" -> recorder.resume(result)
      "isPaused" -> recorder.isPaused(result)
      "isRecording" -> recorder.isRecording(result)
      "cancel" -> recorder.cancel(result)
      "hasPermission" -> hasPermission(call, result)
      "getAmplitude" -> recorder.getAmplitude(result)
      "listInputDevices" -> result.success(DeviceUtils.listInputDevicesAsMap(appContext))
      "dispose" -> disposeRecorder(recorder, recorderId, result)
      "isEncoderSupported" -> isEncoderSupported(call, result)
      else -> result.notImplemented()
    }
  }

  fun dispose() {
    for (entry in recorders.entries) {
      disposeRecorder(entry.value, entry.key, null)
    }

    recorders.clear()
  }

  private fun createRecorder(recorderId: String, result: MethodChannel.Result) {
    try {
      val recorder = RecorderWrapper(appContext, recorderId, messenger)
      recorders[recorderId] = recorder
      result.success(null)
    } catch (e: Exception) {
      result.error("record", "Cannot create recorder.", e.message)
    }
  }

  private fun disposeRecorder(recorder: RecorderWrapper, recorderId: String, result: MethodChannel.Result?) {
    recorder.dispose()
    recorders.remove(recorderId)
    result?.success(null)
  }

  private fun hasPermission(call: MethodCall, result: MethodChannel.Result) {
    val request = call.argument<Boolean>("request") ?: true
    permissionManager.hasPermission(request, result::success)
  }

  private fun isEncoderSupported(call: MethodCall, result: MethodChannel.Result) {
    val codec = call.argument<String>("encoder")

    val isSupported = AudioFormats.isEncoderSupported(
      AudioFormats.getMimeType(Objects.requireNonNull(codec))
    )

    result.success(isSupported)
  }
}